src/components/select/index.tsx

```diff
@@ -0,0 +1,247 @@
+import { Fragment, useCallback, useMemo } from "react";
+import { Listbox, Transition } from "@headlessui/react";
+import { HiCheck, HiChevronDown } from "react-icons/hi";
+
+import Search from "@/components/Search";
+import cn from "@/lib/classnames";
+import isBlank from "@/lib/is-blank";
+
+type SelectValueType<
+  Value,
+  Option,
+  Mutiple extends boolean,
+  UseOption extends boolean
+> = Mutiple extends true
+  ? UseOption extends true
+    ? Option[]
+    : Value[]
+  : UseOption extends true
+    ? Option
+    : Value;
+
+export interface SelectProps<
+  Option extends Record<NameKey | ValueKey, unknown>,
+  Multiple extends boolean = false,
+  UseOption extends boolean = false,
+  NameKey extends string = "name",
+  ValueKey extends string = "value",
+  SelectValue = SelectValueType<Option[ValueKey], Option, Multiple, UseOption>
+> {
+  label?: string;
+  multiple?: Multiple;
+  /** if true, the value will be option instead of value of option */
+  useOption?: UseOption;
+  options?: Option[];
+  value?: SelectValue;
+  onChange?: (value: SelectValue) => void;
+  width?: string;
+  height?: string;
+  placeholder?: string;
+  renderSelected?: (
+    values: SelectValue,
+    options: Option[],
+    onChange?: (value: SelectValue) => void
+  ) => JSX.Element;
+  renderOption?: (option: Option) => JSX.Element;
+  searchValue?: string;
+  onSearch?: (search: string) => void;
+  nameKey?: NameKey;
+  valueKey?: ValueKey;
+  renderList?: (options: Option[]) => JSX.Element;
+  /** When use select as filter
+   * - Do not use with search and dynamic loading, could cause unexpected behaviour
+   */
+  treatEmptyAsAll?: boolean; // todo: support search and dynamic loading senario
+}
+
+/**
+ * Select component (Nonstable)
+ * test page: http://localhost:3000/tests/select_component
+ * ref: https://headlessui.com/react/listbox
+ * todo: support error state
+ * todo: merge with FieldSelct
+ * todo: optimise animation
+ */
+export default function Select<
+  Option extends Record<NameKey | ValueKey, unknown>,
+  Multiple extends boolean = false,
+  UseOption extends boolean = false,
+  NameKey extends string = "name",
+  ValueKey extends string = "value",
+  SelectValue = SelectValueType<Option[ValueKey], Option, Multiple, UseOption>
+>(
+  props: SelectProps<
+    Option,
+    Multiple,
+    UseOption,
+    NameKey,
+    ValueKey,
+    SelectValue
+  >
+) {
+  const {
+    label,
+    multiple = false,
+    useOption = false,
+    options = [],
+    value = [] as SelectValue,
+    onChange,
+    width = "w-full",
```

Comment:
I know you're following the convention of the other components, but this can be simplified by merging width and height into `className` allowing you to pass any stykes as a prop